Fix calls to setContentLength(). (Fixes #197)

dopplershift opened the pull request:
setContentLength() sets the content-length HTTP header, which is defined
as the number of bytes. In all of these changed places, we were passing
in the length of the string, which is the number of characters. This
breaks in the presence UTF-8 characters, which may be encoded as up to 4
bytes; the result is truncated HTML data (observed in #197).

Original report saw stray characters on HTML view; I was not able to see these in chrome, but they were clear (only `</h` of the closing `</html>` tag) when viewing the HTML source. Two items that may warrant discussion:
- I'd like to add a test, but I'm not sure how best to get a useful one that checks the HTML returned when serving up an appropriate catalog. Pointers welcome.
- I hit all of the users of `setContentLength()`, but I'm not sure if that's all of the problem. Anyone know of anything else easily searchable that could be a victim of this?

DennisHeimbigner commented:
Alternately, we could disallow setContentLength();
Is there anywhere that requires it?
=Dennis

Ryan May wrote:
> setContentLength() sets the content-length HTTP header, which is defined
> as the number of bytes. In all of these changed places, we were passing
> in the length of the string, which is the number of characters. This
> breaks in the presence UTF-8 characters, which may be encoded as up to 4
> bytes; the result is truncated HTML data (observed in #197).
> 
> Original report saw stray characters on HTML view; I was not able to see these in chrome, but they were clear (only `</h` of the closing `</html>` tag) when viewing the HTML source. Two items that may warrant discussion:
> - I'd like to add a test, but I'm not sure how best to get a useful one that checks the HTML returned when serving up an appropriate catalog. Pointers welcome.
> - I hit all of the users of `setContentLength()`, but I'm not sure if that's all of the problem. Anyone know of anything else easily searchable that could be a victim of this?
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/Unidata/thredds/pull/271
> 
> -- Commit Summary --
> 
>   * Fix calls to setContentLength(). (Fixes #197)
> 
> -- File Changes --
> 
>     M tds/src/main/java/thredds/core/ConfigCatalogHtmlWriter.java (8)
>     M tds/src/main/java/thredds/server/cdmremote/CdmRemoteController.java (3)
>     M tds/src/main/java/thredds/server/cdmremote/CdmrFeatureController.java (2)
>     M tds/src/main/java/thredds/server/cdmvalidator/CdmValidatorController.java (4)
>     M tds/src/main/java/thredds/server/metadata/MetadataController.java (2)
>     M tds/src/main/java/thredds/server/ncss/controller/NcssDatasetInfoController.java (2)
>     M tds/src/main/java/thredds/servlet/DataRootHandler.java (2)
>     M tds/src/main/java/thredds/servlet/HtmlWriter.java (15)
>     M tds/src/main/java/thredds/servlet/ServletUtil.java (14)
> 
> -- Patch Links --
> 
> https://github.com/Unidata/thredds/pull/271.patch
> https://github.com/Unidata/thredds/pull/271.diff
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Unidata/thredds/pull/271
>

dopplershift commented:
No idea as to the utility of setting the value. I can say that in a couple places we return this size out of our API (listed as "bytes"), so those still need to be corrected. Not sure what the ramifications are of that value being incorrect.

cwardgar commented:
For the test, you could check that the returned HTML is well-formed. That, at least, would catch the closing tag being `<\h` instead of `</html>`.

dopplershift commented:
For the test, I'm more clueless as the right layer to come in and how to set up testing a server response. Can I mock if I really want to test the actual server functionality?

cwardgar commented:
Do you want to test the behavior of a Controller? If so, yes, you can completely mock that interaction using the `spring-test` stuff. `AnyHtmlControllerTest` is a good example. Also, maybe you just want to test `ConfigCatalogHtmlWriter.writeCatalog()`? In that case, you could probably avoid the `WebAppConfiguration` stuff (since Spring isn't autowiring a controller for you) and just use `MockHttpServletRequest` and `MockHttpServletResponse`. You'll still probably want a real `Catalog` though.

Some related weirdness that I noticed: in 4.6 `ConfigCatalogHtmlWriter.writeCatalog()` is only being called from 2 unused controllers: `LocalCatalogServiceController2` and `RemoteCatalogServiceController2`. In 5.0, it's being called from `CatalogServiceController` and `RemoteCatalogServiceController`. So is it unused in 4.6? Why are we seeing the behavior in #197?

DennisHeimbigner commented:
Remember that the body of the message may not ascii or even utf8;
it could be binary, in which case the specified content length may be correct.

DennisHeimbigner commented:
In looking at some of the changes, we probably need to do something like this:
    X.setContentLength(string.getBytes(Charset.forName("UTF-8")).length())
at least for those cases where we have a string.

dopplershift commented:
Can you point me to a place I've changed where binary data may be present? As far as I can tell all places were either some version of a catalog or a place where it was explicitly converting something to a string already.

As far as this:
'''
In looking at some of the changes, we probably need to do something like this:
X.setContentLength(string.getBytes(Charset.forName("UTF-8")).length())
at least for those cases where we have a string.
'''
How exactly is that different from what I added in `ServletUtils.java`:
```java
   /**
* Set the proper content length for the string
*
* @param response the HttpServletResponse to act upon
* @param s the string that will be returned
* @return the number of bytes
* @throws UnsupportedEncodingException on bad character encoding
*/
public static int setResponseContentLength(HttpServletResponse response, String s) throws UnsupportedEncodingException {
 int length = s.getBytes(response.getCharacterEncoding()).length;
 response.setContentLength(length);
 return length;
}

```

DennisHeimbigner commented:
I must have been looking at the wrong thing,
because I thought I still saw things of the form
...setContentLength(s.length()).


Ryan May wrote:
> Can you point me to a place I've changed where binary data may be present? As far as I can tell all places were either some version of a catalog or a place where it was explicitly converting something to a string already.
> 
> As far as this:
> '''
> In looking at some of the changes, we probably need to do something like this:
> X.setContentLength(string.getBytes(Charset.forName("UTF-8")).length())
> at least for those cases where we have a string.
> '''
> How exactly is that different from what I added in `ServletUtils.java`:
> ```java
>    /**
> * Set the proper content length for the string
> *
> * @param response the HttpServletResponse to act upon
> * @param s the string that will be returned
> * @return the number of bytes
> * @throws UnsupportedEncodingException on bad character encoding
> */
> public static int setResponseContentLength(HttpServletResponse response, String s) throws UnsupportedEncodingException {
>  int length = s.getBytes(response.getCharacterEncoding()).length;
>  response.setContentLength(length);
>  return length;
> }
> 
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Unidata/thredds/pull/271#issuecomment-154463339

JohnLCaron commented:
Setting reponse.setContentLength() gets the length into the HTTP header, and into threddsServlet.log. For a while I was trying to use it when possible.

If not set, i _think_ tomcat may set length if content is small enough to fit into buffer. otherwise, it cant set it (because the header must go before the content, who designed that?) so does a chunked-encoding thing and streams out the content.

Anyway, probably not very important to make it work (but thanks for getting it right when we do set it!) because the size does get set in tomcat's access.log, obviously written after the response is sent. perhaps i didnt realize that at the time i worked to get that number into threddsServlet.logs. perhaps makes client job easier in some cases if it knows in advance?

dopplershift commented:
This should be ready to merge now.
